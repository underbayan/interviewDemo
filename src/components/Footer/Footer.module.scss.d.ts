export const footerStyle: string;
export const infoStyle: string;
