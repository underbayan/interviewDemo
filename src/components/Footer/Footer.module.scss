@import "@demo/designSystem";

.footerStyle {
  display: flex;
  width: 100vw;
  align-items: center;
  justify-content: center;
  position: fixed;
  bottom: 0;
  font-size: 11px;
  row-gap: 2px;
  flex-direction: column;
  border-top: 1px solid var(--grays-5-color);
  background-color: var(--grays-7-color);

  .infoStyle {
    flex: 1;
  }
}