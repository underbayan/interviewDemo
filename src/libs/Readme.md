"Libs" refer to the libraries or modules that contain code that is meant to be reusable across different parts of an application. In the context of the specified constraint, the code in "libs" should not contain any mutations logic. Rather, it should be focused on providing functionalities that are shared across different parts of the application without altering any data. For instance, "libs" could contain helper functions, utility classes, or any other non-mutating code that can be used in different parts of the application.