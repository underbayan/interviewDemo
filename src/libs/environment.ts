import stringsEn from "../mock.strings.en.json"
import stringsZh from "../mock.strings.zh.json"

export const environment = {
  app: "demo",
  page: "/",
  lang: "en",
  version: "0.0.1",
  flights: "", // AB test
  config: {},  // Config for page
  strings: stringsEn,
  langMap: {
    en: stringsEn,
    zh: stringsZh
  }
  // TODO: i18n.detectLanguage() to get zh/en/fr lang config json
}