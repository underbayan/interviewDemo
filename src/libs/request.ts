export interface FetchDataType {
  [key: string]: any
}

export interface AbortableFetchResponse extends Response {
  data?: FetchDataType | string | ArrayBuffer | Blob
  message: string
}

export interface AbortableFetchOptions extends RequestInit {
  abort?: AbortSignal
}

export interface AbortableFetchPromise extends Promise<AbortableFetchResponse> {
  [AbortableKey](): void
}

export const AbortableKey = Symbol.for('abortableKey')
export const abortableFetch = (
  url: string,
  options?: AbortableFetchOptions
): AbortableFetchPromise => {
  const { abort, ...fetchOptions } = options ?? {}
  const ab = new AbortController()
  const rt = (fetch(url, { ...fetchOptions, signal: abort ? abort : ab.signal }) as Promise<Response>)
    .then(async (response) => {
      const contentType = response.headers.get('content-type')
      let data: FetchDataType | undefined | string

      if (contentType?.includes('application/json')) {
        data = await response.json()
      } else if (contentType?.includes('application/text')) {
        data = await response.text()
      } else if (contentType?.includes('application/opaque')) {
        data = await response.blob()
      } else {
        data = await response.arrayBuffer()
      }

      const result: AbortableFetchResponse = {
        ...response,
        status: response.status,
        message: response.statusText,
        data,
      }
      return result
    })
    .catch((e) => console.error(e)) as AbortableFetchPromise

  Object.defineProperty(rt, AbortableKey, { value: () => ab.abort(), writable: false })

  return rt
}
