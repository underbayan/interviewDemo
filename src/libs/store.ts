import { createSlice, configureStore } from '@reduxjs/toolkit'
import { useSelector, useDispatch } from 'react-redux'
import { environment } from "./environment"
import { Dispatch, AnyAction } from 'redux'
export type RootState = ReturnType<typeof store.getState>

const envSlice = createSlice({
  name: 'env',
  initialState: environment,
  reducers: {

    updateEnv(state, action) {

      const newState = { ...state, ...action?.payload }
      if (action?.payload?.lang) newState.strings = newState.langMap[action.payload?.lang]
      return newState
    }
  }
})
export const store = configureStore({
  reducer: {
    env: envSlice.reducer
  },
})
interface EnvUpdatePayload {
  lang?: string
  [key: string]: any
}




export const useI18n = () => useSelector((state: RootState) => state.env.strings)
export const useEnv = () => useSelector((state: RootState) => state.env)
export const useUpdateEnv = (): ((env: EnvUpdatePayload) => void) => {
  const dispatch = useDispatch()

  const updateEnvAsync = (env: EnvUpdatePayload) => (dispatch: Dispatch) => {
    setTimeout(() => { dispatch(envSlice.actions.updateEnv(env)) }, 15)
  }

  return (env: EnvUpdatePayload) => dispatch(updateEnvAsync(env) as unknown as AnyAction)
}


