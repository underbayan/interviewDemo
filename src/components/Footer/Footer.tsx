import { footerStyle, infoStyle } from "./Footer.module.scss"
import { useI18n } from "../../libs/store"

export const Footer = () => {
  const str = useI18n()
  return (
    <footer className={"p-xs-1 p-sm-2 " + footerStyle}>
      <div className={infoStyle}>{str.footerAuthor}</div>
      <div className={infoStyle}>{str.footerCopyright.replace("{date}", new Date().getFullYear().toString())}</div>
    </footer>
  )
}