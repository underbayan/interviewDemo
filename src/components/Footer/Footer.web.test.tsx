import { render } from '@testing-library/react';
import { test, expect } from 'vitest';
import { Footer } from './Footer';
import { store } from "@demo/libs/store"
import { Provider } from 'react-redux'

test('Footer renders correctly', () => {
  const { container } = render(<Provider store={store}>
    <Footer>Click me</Footer>
  </Provider>);
  expect(container).toMatchSnapshot();
});
