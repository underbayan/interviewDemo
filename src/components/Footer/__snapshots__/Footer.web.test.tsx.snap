// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer renders correctly 1`] = `
<div>
  <footer
    class="p-xs-1 p-sm-2 undefined"
  >
    <div>
      Made with ❤️ in Melbourne.
    </div>
    <div>
      @ 2023 Broccoli & Co. All rights reserved.
    </div>
  </footer>
</div>
`;
